src/main/scala/com/fulcrumgenomics/sv/tools/AggregateSvPileup.scala

```diff
@@ -0,0 +1,319 @@
+import com.fulcrumgenomics.util.{Io, Metric}
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+
+/** Type aliases */
+trait AggregateSvPileupDef {
+  type PileupGroup = Seq[BreakpointPileup]
+}
+
+
+@clp(group=ClpGroups.All, description=
+  """
+    |Merges nearby pileups of reads supporting putative structural variants.
+    |
+    |Takes as input the file of pileups produced by `SvPileup`. That file contains a list of breakpoints, each
+    |consisting of a chromosome, position and strand for each side of the breakpoint, as well as quantified read support
+    |for the breakpoint.
+    |
+    |This tool merges pairs of breakpoints that have their left and right ends on the same chromosomes, the same left
+    |and right strands, and their left and right positions both within a length threshold. The merging behavior is
+    |transitive. For example, two breakpoints that are farther apart than the length threshold can be merged if there
+    |is a third breakpoint that is close to both.
+    |
+    |The output file is a tab-delimited table with one record per aggregated cluster of pileups. Aggregated
+    |pileups are reported with the minimum and maximum (inclusive) coordinates of all pileups in the cluster, a
+    |possible putative structural variant event type supported by the pileups, and the sum of read support from all
+    |pileups in the cluster.
+    |""")
+class AggregateSvPileup
+(@arg(flag='i', doc="The input text file of pileups generated by SvPileup") input: FilePath,
+ @arg(flag='o', doc="The output file") output: FilePath,
+ @arg(flag='d', doc="Distance threshold below which to cluster breakpoints") maxDist: Int = 100,
+) extends SvTool with AggregateSvPileupDef {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    // Read pileups from input file
+    val pileups: Seq[BreakpointPileup] = Metric.read[BreakpointPileup](input)
+
+    // Group pileups by left and right contig and strand
+    val pileupsByContigStrand: Map[BreakpointContigOrientation, PileupGroup] =
+      pileups.groupMap(BreakpointContigOrientation(_))(identity)
+
+    // Within each contig+strand group, identify all immediate "neighbors" of each pileup (other pileups within the
+    // distance threshold)
+    val pileupToNeighbors: Map[BreakpointContigOrientation, Map[BreakpointPileup, PileupGroup]] =
+      pileupsByContigStrand.view.mapValues(AggregateSvPileup.pileupToNeighbors(_, maxDist)).toMap
+
+    // Within each contig+strand group, recursively build (transitive) clusters of neighboring pileups
+    val pileupClusters: Map[BreakpointContigOrientation, Seq[PileupGroup]] =
+      pileupToNeighbors.view.mapValues(AggregateSvPileup.toClusters).toMap
+
+    // Aggregate each cluster
+    val aggregatedPileups: Map[BreakpointContigOrientation, Seq[AggregatedBreakpointPileup]] =
+      pileupClusters.view.mapValues(_.map(AggregatedBreakpointPileup.apply)).toMap
```

Comment:
I am having a really hard time reviewing and following along `toClusters` and friends for what I think should be a really straightforward BFS or DFS search.

If I understand correctly, the algorithm is:

1. group by the same left/right contig+strand
2. for each group, create an adjacency matrix
3. find each connected component in each adjacency matrix
4. for each connected component, create an `AggregatedBreakpointPileup`
5. output

If so, I think the code can be simplified such that after creating the adjacency matrix, we perform a BFS or DFS search within each pileup group

1. Keep a boolean to say if a breakpoint has been "visited", default to false
2. Loop through each breakpoint.  If visited, do nothing.  Else it has not been visited, so add it to the current "cluster", mark as visited, and recurse on all other adjacent breakpoints adding them to the current cluster.
3. Done

Reply:
I'm having trouble designing this algorithm in a way that doesn't require an `extractClusterFor()` method to extract the cluster containing a given representative pileup. Let's discuss live.